napari/plugins/utils.py

```diff
@@ -0,0 +1,179 @@
+    instances in the corresponding hook caller) and reorders the implementations in the hook
+    caller accordingly.
+
+    Note: hook implementations are actually stored in *two* separate list attributes in the
+    hook caller: ``_HookCaller._wrappers`` and ``_HookCaller._nonwrappers``, according to
+    whether the corresponding ``HookImpl`` instance was marked as a wrapper or not.  For
+    more, see: https://pluggy.readthedocs.io/en/latest/#wrappers
+
+    Parameters
+    ----------
+    hook_caller : pluggy.hooks._HookCaller
+        The hook caller to reorder
+    order : list
+        A list of str (plugin names), or ``HookImpl`` instances, in the desired CALL ORDER
+        of the hook implementations.  The list does not *need* to include every hook
+        implementation in the ``hook_caller.get_hookimpls()``, but those that are not
+        included will be moved to the end of the call order.
+
+    Raises
+    ------
+    TypeError
+        If ``order`` is neither entirely strings nor entirely HookImpl
+        instances.
+    ValueError
+        If the 'order' list cannot be interpreted as a list of "plugin_name"
+        or ``HookImpl`` instances.
+    ValueError
+        if 'order' argument has multiple entries for the same implementation.
+
+    Examples
+    --------
+    Imagine you had a hook specification named ``print_plugin_name``, that expected plugins
+    to simply print their own name. An implementation might look like:
+
+    >>> # hook implementation for ``plugin_1``
+    >>> @hook_implementation
+    ... def print_plugin_name():
+    ...     print("plugin_1")
+
+    If three different plugins provided hook implementations. An example call for that hook
+    might look like:
+
+    >>> plugin_manager.hook.print_plugin_name()
+    plugin_1
+    plugin_2
+    plugin_3
+
+    If you wanted to rearrange their call order, you could do this:
+
+    >>> hook_caller = plugin_manager.hook.print_plugin_name
+    >>> new_order = ["plugin_2", "plugin_3", "plugin_1"]
+    >>> permute_hook_implementations(hook_caller, new_order)
+    >>> plugin_manager.hook.print_plugin_name()
+    plugin_2
+    plugin_3
+    plugin_1
+
+    You can also just specify one or more item to move them to the front of the call order:
+    >>> permute_hook_implementations(hook_caller, ["plugin_3"])
+    >>> plugin_manager.hook.print_plugin_name()
+    plugin_3
+    plugin_2
+    plugin_1
+    """
+    # make sure items in order are unique
+    if len(order) != len(set(order)):
+        raise ValueError("repeated item in order")
+
+    # get a list of all hook implementations in hook_caller
+    hook_implementations = hook_caller.get_hookimpls()
+    if len(order) > len(hook_implementations):
+        raise ValueError(
+            "too many values in order: "
+            f"({len(order)} > {len(hook_implementations)})"
+        )
+
+    # figure out what list type was entered for ``order`` (or raise error)
+    order_type = _guess_order_type(order)
```

Comment:
I'd call this `_validate_order_type`.

But, to be honest, why not allow both types arbitrarily, but homogenise to the actual objects, then you can forget about the names below?

Reply:
gone now